plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/TransformerHelpers.kt

```diff
@@ -0,0 +1,144 @@
+import org.jetbrains.kotlin.types.Variance
+import kotlin.random.Random
+
+interface TransformerHelpers {
+
+    val context: IrPluginContext
+
+    fun buildCall(
+        target: IrFunctionSymbol,
+        type: IrType? = null,
+        origin: IrStatementOrigin,
+        typeArguments: List<IrType>? = null,
+        valueArguments: List<IrExpression?>? = null
+    ): IrCall =
+        IrCallImpl(
+            UNDEFINED_OFFSET,
+            UNDEFINED_OFFSET,
+            type ?: target.owner.returnType,
+            target,
+            target.descriptor.typeParametersCount,
+            origin
+        ).apply {
+            typeArguments?.let {
+                assert(typeArguments.size == typeArgumentsCount)
+                it.withIndex().forEach { (i, t) -> putTypeArgument(i, t) }
+            }
+            valueArguments?.let {
+                assert(valueArguments.size == valueArgumentsCount)
+                it.withIndex().forEach { (i, arg) -> putValueArgument(i, arg) }
+            }
+        }
+
+    fun buildSetField(
+        symbol: IrFieldSymbol,
+        receiver: IrExpression?,
+        value: IrExpression,
+        superQualifierSymbol: IrClassSymbol? = null
+    ) =
+        IrSetFieldImpl(
+            UNDEFINED_OFFSET,
+            UNDEFINED_OFFSET,
+            symbol,
+            receiver,
+            value,
+            value.type,
+            IrStatementOrigin.INVOKE,
+            superQualifierSymbol
+        )
+
+    fun buildGetField(symbol: IrFieldSymbol, receiver: IrExpression?, superQualifierSymbol: IrClassSymbol? = null, type: IrType? = null) =
+        IrGetFieldImpl(
+            UNDEFINED_OFFSET,
+            UNDEFINED_OFFSET,
+            symbol,
+            type ?: symbol.owner.type,
+            receiver,
+            IrStatementOrigin.GET_PROPERTY,
+            superQualifierSymbol
+        )
+
+    fun buildFunctionSimpleType(typeParameters: List<IrType>): IrSimpleType {
+        val parametersCount = typeParameters.size - 1
+        val classDesc = context.irBuiltIns.builtIns.getFunction(parametersCount)
+        val symbol = context.symbolTable.referenceClass(classDesc)
+        return IrSimpleTypeImpl(
+            classifier = symbol,
+            hasQuestionMark = false,
+            arguments = typeParameters.map { it.toTypeArgument() },
+            annotations = emptyList()
+        )
+    }
+
+    fun getterName(name: String) = "<get-$name>${Random.nextInt()}"
+    fun setterName(name: String) = "<set-$name>${Random.nextInt()}"
+    fun Name.getFieldName() = "<get-(\\w+)>".toRegex().find(asString())?.groupValues?.get(1)
+
+    private fun IrType.toTypeArgument(): IrTypeArgument {
+        return makeTypeProjection(this, Variance.INVARIANT)
+    }
+
+    fun IrCall.getValueArguments() = (0 until valueArgumentsCount).map { i ->
+        getValueArgument(i)
+    }
+
+    fun IrValueParameter.capture() = JsIrBuilder.buildGetValue(symbol)
+
+    fun referencePackageFunction(
```

Comment:
Are the following methods reference external symbols? If so, please use corresponding `IrPluginContext` API